src/server/game/Entities/AreaTrigger/AreaTrigger.cpp

```diff
@@ -642,6 +657,71 @@ bool AreaTrigger::HasSplines() const
     return bool(_spline);
 }
 
+void AreaTrigger::InitCircularMovement(AreaTriggerCircularMovementInfo const& /*cmi*/, uint32 timeToTarget)
+{
+    // should be sent in object create packets only
+    m_uint32Values[AREATRIGGER_TIME_TO_TARGET] = timeToTarget;
+
+    if (IsInWorld())
+    {
+        // TODO: broadcast AreaTriggerReShape?
+    }
+}
+
+AreaTriggerCircularMovementInfo AreaTrigger::GetAreaTriggerCircularMovementInfo() const
+{
+    AreaTriggerCircularMovementInfo areaTriggerCircularMovementInfo;
+    AreaTriggerMiscTemplate const* areaTriggerMiscTemplate = GetMiscTemplate();
+
+    areaTriggerCircularMovementInfo = areaTriggerMiscTemplate->CircularMovementInfo;
+
+    areaTriggerCircularMovementInfo.TimeToTarget = GetTimeToTarget();
+    areaTriggerCircularMovementInfo.ElapsedTimeForMovement = GetElapsedTimeForMovement();
+
+    if (!_circularMovementCenterGUID.IsEmpty())
+        areaTriggerCircularMovementInfo.TargetGUID = _circularMovementCenterGUID;
+
+    if (!_circularMovementCenterPosition.IsPositionEmpty())
+        areaTriggerCircularMovementInfo.Center = _circularMovementCenterPosition;
+
+    return areaTriggerCircularMovementInfo;
+}
+
+Position const& AreaTrigger::GetCircularMovementCenterPosition() const
+{
+    if (_circularMovementCenterPosition.IsPositionEmpty() && !_circularMovementCenterGUID.IsEmpty())
+        if (WorldObject* center = ObjectAccessor::GetWorldObject(*this, _circularMovementCenterGUID))
+            return *center;
+
+    return _circularMovementCenterPosition;
+}
+
+void AreaTrigger::UpdateCircularMovementPosition()
+{
+    Position const& centerPos = GetCircularMovementCenterPosition();
+    AreaTriggerCircularMovementInfo const& cmi = GetMiscTemplate()->CircularMovementInfo;
+
+    // AreaTrigger make exactly "TimeToTarget / Duration" loops during his life time
+    float angleDiff = 2.0f * float(M_PI) / (float(GetTimeToTarget()) / float(GetDuration())) * (float(GetTimeSinceCreated()) / float(GetDuration()));
```

Comment:
it is the same like `2.0f * float(M_PI) * (float(GetTimeSinceCreated()) / float(GetTimeToTarget()))`